envpool/sokoban/sokoban_node.cc

```diff
@@ -0,0 +1,196 @@
+// Copyright 2023-2024 FAR AI
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "envpool/sokoban/sokoban_node.h"
+
+#include <algorithm>
+#include <limits>
+
+namespace sokoban {
+
+const std::vector<std::pair<int, int>> SokobanNode::kDelta = {
+    {0, -1}, {0, 1}, {-1, 0}, {1, 0}  // Up, Down, Left, Right
+};
+
+bool SokobanNode::IsSameState(SokobanNode& rhs) {
+  if (player_x != rhs.player_x || player_y != rhs.player_y) {
+    return false;
+  }
+  for (size_t i = 0; i < boxes.size(); i++) {
```

Comment:
range-based iter

Reply:
Here since we iterate through both boxes & rhs.boxes, iterating with int i seems best, unless there's another alternative I'm not aware of.

Reply:
oh yeah you're right, I thought it was a double loop. Nevermind!

Reply:
I think the best thing here is to use `boxes != rhs.boxes`.

Reply:
Aah that's right. Somehow I keep forgetting that comparison operators are overloaded for many such classes.